File: C++/combination-sum.cpp
//link to problem-https://leetcode.com/problems/combination-sum/


class Solution {
public:
    
  
    vector<int>v1;
    
    void find(int ind,int n,vector<int>& candidates,int target,vector<vector<int>>&v)
    {
        if(ind==n)
        {
         if(target==0)
         
            v.push_back(v1);
         
            return ;
         }
        
        if(target-candidates[ind]>=0)
        {
            //pick
            v1.push_back(candidates[ind]);
            find(ind,n,candidates,target-candidates[ind],v);
            v1.pop_back();
        }
        
        find(ind+1,n,candidates,target,v);//non pick
        
    }
    
    
    
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
  vector<vector<int>> v;
        
        find(0,candidates.size(), candidates,target, v);
        return v;
    }
};

File: Java/Combination_Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans=new ArrayList<>();
        findCombination(0,candidates,target,ans,new ArrayList<>());
        return ans;
    }
    void findCombination(int index,int[] arr,int target,List<List<Integer>> ans,List<Integer> ds){
        if(index==arr.length){
            if(target==0){
                ans.add(new ArrayList<>(ds));
            }
            return;
        }
        if(arr[index]<=target){
            ds.add(arr[index]);
             findCombination(index,arr,target-arr[index],ans,ds);
            ds.remove(ds.size()-1);
        }
            findCombination(index+1,arr,target,ans,ds);
        
    }
}
